apps/admin_app/assets/js/components/CreatePromotions.tsx

```diff
@@ -0,0 +1,369 @@
+            method:'POST',
+            body:JSON.stringify(data),
+            headers: {'Content-Type': 'application/json; charset=UTF-8'},
+        }).then(res =>res.json())
+        
+    }
+
+    fetchGET =(url)=>{
+        return fetch(url,{
+            credentials:'include',
+            method:"GET",
+            headers:{'Content-Type': 'application/json; charset=UTF-8'}
+        }).then(res =>res.json())
+    }
+
+    handleSave = () =>{
+        var data = {
+            "data":{
+                "starta_at":this.state.starts_at,
+                "expires_at":this.state.expires_at,
+                "code":this.state.code,
+                "name":this.state.name,
+                "rules":[],
+                "actions":[]
+            }
+        }
+        var url = 'http://localhost:4000/api/promotions'
+        this.fetchPOST(url,data).then(response=>{
+            var arrofkeys=Object.keys(response)
+            arrofkeys[0]=="errors"?this.onFail(response):this.onSuccess()
+            console.log('Success',JSON.stringify(response))})
+        .catch(error=>console.error('Error',error));
+        
+    }
+
+    onSuccess =()=>{
+        this.setState({success:true})
+        alert("Saved Successfully!")
+    }
+    onFail=(response)=>{
+        Object.keys(response).map((res)=>{
+            var errormsg=""
+            Object.keys(response[res]).map((res1)=>{
+                errormsg=errormsg+res1+" "+response[res][res1][0]["message"]+"\n"
+            })
+            alert(errormsg)
+        })
+    }
+    handleAddRule=()=>{
+        
+        this.setState({addRule:true})
+        var url="http://localhost:4000/api/promo-rules"
+        this.fetchGET(url).then(response=>{
+            var availableRules=response
+            this.setState({availableRules:availableRules})
+            console.log('Success',JSON.stringify(response))})
+        .catch(error=>console.error('Error',error));
+        
+    }
+
+    handleAddAction=()=>{
+        this.setState(prevState=>({
+            actionDown:!prevState.actionDown,
+            addAction:!prevState.addAction
+        }))
+
+        var url="http://localhost:4000/api/promo-actions"
+        this.fetchGET(url).then(response=>{
+            var availableActions=response
+            this.setState({availableActions:availableActions})
+            console.log('Success',JSON.stringify(response))})
+        .catch(error=>console.error('Error',error));
+        
+        this.fetchGET("http://localhost:4000/api/promo-calculators").then(response=>{
+            var availableCalculators=response
+            this.setState({availableCalculators:availableCalculators})
+            console.log('Success',JSON.stringify(response))})
+        .catch(error=>console.error('Error',error));
+    }
+
+    handleRule=(rule)=>{
+        console.log("module",rule)
+        this.setState({selectedRule:rule})
+        var url="http://localhost:4000/api/promo-rule-prefs"
```

Comment:
using var instead of const and hardcoded url - Not ideal way.
Put this url inside some constants file and then use here.